src/main/scala/mesosphere/marathon/core/instance/Instance.scala

```diff
@@ -32,18 +32,56 @@ case class Instance(
     unreachableStrategy: UnreachableStrategy,
     reservation: Option[Reservation]) extends MarathonState[Protos.Json, Instance] with Placed {
 
+  def phase(now: Timestamp): InstancePhase = {
+    val maybeHighestPriorityState = InstanceState.conditionFromTasks(tasksMap.values, now, unreachableStrategy)
+
+    if (maybeHighestPriorityState.exists(s => s == Condition.Created || s == Condition.Staging || s == Condition.Starting)) {
+      InstancePhase.Launching
+    } else if (maybeHighestPriorityState.exists(s => s == Condition.Running || s == Condition.Unreachable || s == Condition.Killing)) {
+      InstancePhase.Active
+    } else if (state.goal == Goal.Running) {
+      InstancePhase.Scheduled
+    } else {
+      InstancePhase.Terminal
+    }
+  }
+
+  /**
+    * This picks
+    * Previously [[Condition]] was used as a status for both tasks and instances.
+    * We moved away from using that for instances but it's still needed for backward compatibility on API layer.
+    * @param now current time
+    * @return computed [[Condition]] as a summary of task conditions
+    */
+  @deprecated("This should be used only in the API layer for backward compatibility and internally inside instance")
+  def summarizedTaskStatus(now: Timestamp): Condition = {
+    InstanceState.conditionFromTasks(tasksMap.values, now, unreachableStrategy).getOrElse(Condition.Unknown)
+  }
+
   val runSpecId: PathId = instanceId.runSpecId
 
   // An instance has to be considered as Reserved if at least one of its tasks is Reserved.
-  def isReserved: Boolean = tasksMap.values.exists(_.status.condition == Condition.Reserved)
-
-  def isReservedTerminal: Boolean = tasksMap.values.exists(_.isReservedTerminal)
+  lazy val isReserved: Boolean = tasksMap.valuesIterator.exists(_.status.condition == Condition.Reserved)
+  lazy val isReservedTerminal: Boolean = tasksMap.valuesIterator.exists(_.isReservedTerminal)
+
+  // these state cannot change without the instance itself being changed and it's immutable so we are safe to use lazy val
+  lazy val isStaging: Boolean = summarizedTaskStatus(Timestamp.now()) == Condition.Staging
```

Comment:
Hm, I'm not sure we should use `Timestamp.now()` in combination with a `lazy val`.